query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRestClient.java

```diff
@@ -0,0 +1,114 @@
+package org.hypertrace.core.query.service.prometheus;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import lombok.NoArgsConstructor;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+public class PrometheusRestClient {
+  private static final String INSTANT_QUERY = "api/v1/query";
+  private static final String RANGE_QUERY = "api/v1/query_range";
+
+  private String host;
+  private int port;
+  private OkHttpClient okHttpClient;
+
+  public PrometheusRestClient(String host, int port, OkHttpClient okHttpClient) {
+    this.okHttpClient = okHttpClient;
+    this.host = host;
+    this.port = port;
+  }
+
+  @SuppressWarnings("unchecked")
+  public List<PrometheusMetricQueryResponse> execute(PromQLQuery query) {
+    List<Request> requests =
+        query.isInstantRequest() ? getInstantQueryRequests(query) : getRangeQueryRequests(query);
+
+    CompletableFuture<Response>[] completableFutures =
+        requests.stream()
+            .map(
+                request -> {
+                  Call call = okHttpClient.newCall(request);
+                  OkHttpResponseCallback callback = new OkHttpResponseCallback();
+                  call.enqueue(callback);
+                  return callback.future;
+                })
+            .toArray(CompletableFuture[]::new);
+
+    CompletableFuture.allOf(completableFutures).join();
```

Comment:
Doing this way, as want to throw an exception if any of the requests fails, would `thanApply` do the same?

Any suggestion here of doing a better way?